src/quadiron-c.cpp

```diff
@@ -0,0 +1,450 @@
+            parities_vec[i] = data[i] + metadata_size;
+            data_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            if (!missing_idxs[fec->n_data + i]) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret = parities_props[fec->n_data + i].from_buffer(
+                    metadata, metadata_size / 4);
+                if (-1 == ret)
+                    return -1;
+            }
+            parities_vec[fec->n_data + i] = parity[i] + metadata_size;
+        }
+    }
+
+    int need_decode = 0;
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        /*
+         * Easy case where the target is a data then simply decode
+         */
+        if (destination_idx < fec->n_data) {
+            wanted_idxs_vec[destination_idx] = 1;
+            res = fec->decode_blocks_vertical(
+                data_vec,
+                parities_vec,
+                parities_props,
+                missing_idxs_vec,
+                wanted_idxs_vec,
+                block_size);
+            if (!res) {
+                ret = -1;
+                goto end;
+            }
+            ret = 0;
+            goto end;
+        }
+    }
+
+    /*
+     * At this point we want a parity to be reconstructed
+     * If systematic we may need to decode if a data is missing.
+     * If non-systematic we always need to decode
+     */
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            if (missing_idxs[i]) {
+                need_decode = 1;
+                wanted_data_vec[i] = 1;
+                data_vec[i] = static_cast<uint8_t*>(malloc(block_size));
+                if (NULL == data_vec[i]) {
+                    ret = -1;
+                    goto end;
+                }
+            }
+        }
+    } else {
+        need_decode = 1;
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            wanted_data_vec[i] = 1;
+            data_vec[i] = static_cast<uint8_t*>(malloc(block_size));
+            if (NULL == data_vec[i]) {
+                ret = -1;
+                goto end;
+            }
+        }
+    }
+
+    if (need_decode) {
+        res = fec->decode_blocks_vertical(
+            data_vec,
+            parities_vec,
+            parities_props,
+            missing_idxs_vec,
+            wanted_data_vec,
+            block_size);
+        if (!res) {
+            ret = -1;
+            goto end;
+        }
+    }
+
+    /*
+     * At this point we have all data blocks.
+     */
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        wanted_idxs_vec[destination_idx - fec->n_data] = 1;
+    } else {
+        wanted_idxs_vec[destination_idx] = 1;
+    }
+
+    fec->encode_blocks_vertical(
+        data_vec, parities_vec, parities_props, wanted_idxs_vec, block_size);
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
```

Comment:
the ~30 following lines look very very similar with the `encode` epilogue.

If possible, it should be factored out into a dedicated helper.